File: ClaudeLight.Python/hook_installer.py
#!/usr/bin/env python3
"""Auto-install hooks into Claude Code settings.json."""
import json
import sys
from pathlib import Path

HOOK_MARKER = "claude-light"

def get_settings_path() -> Path:
    return Path.home() / ".claude" / "settings.json"

def get_cli_path() -> str:
    return str(Path(__file__).parent / "hook_cli.py")

def install_hooks():
    settings_path = get_settings_path()
    cli_path = get_cli_path()
    python_exe = sys.executable

    settings = {}
    if settings_path.exists():
        settings = json.loads(settings_path.read_text(encoding="utf-8"))

    hooks = settings.get("hooks", {})
    if any(HOOK_MARKER in json.dumps(v) for v in hooks.values()):
        print("Hooks already installed.")
        return

    settings["hooks"] = {
        "PreToolUse": [{
            "hooks": [{
                "type": "command",
                "command": f'"{python_exe}" "{cli_path}" hook running $CLAUDE_PROJECT_DIR'
            }]
        }],
        "PostToolUse": [{
            "hooks": [{
                "type": "command",
                "command": f'"{python_exe}" "{cli_path}" hook done $CLAUDE_PROJECT_DIR'
            }]
        }],
        "Notification": [{
            "hooks": [{
                "type": "command",
                "command": f'"{python_exe}" "{cli_path}" hook confirm $CLAUDE_PROJECT_DIR'
            }]
        }]
    }

    settings_path.parent.mkdir(parents=True, exist_ok=True)
    settings_path.write_text(json.dumps(settings, indent=2, ensure_ascii=False), encoding="utf-8")
    print(f"Hooks installed to {settings_path}")

def remove_hooks():
    settings_path = get_settings_path()
    if not settings_path.exists():
        return

    settings = json.loads(settings_path.read_text(encoding="utf-8"))
    if "hooks" in settings:
        del settings["hooks"]
        settings_path.write_text(json.dumps(settings, indent=2, ensure_ascii=False), encoding="utf-8")
        print(f"Removed hooks from {settings_path}")

if __name__ == "__main__":
    if len(sys.argv) > 1 and sys.argv[1] == "remove":
        remove_hooks()
    else:
        install_hooks()

File: ClaudeLight.CSharp/ClaudeLight.CLI/Services/HookInstaller.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text.Json;

namespace ClaudeLight.Services;

public static class HookInstaller
{
    private const string HookMarker = "claude-light";

    public static string EnsureCliInstalled()
    {
        // 直接使用 ClaudeLight.exe 同目录下的 ClaudeLight.CLI.exe
        // 这样 CLI 可以找到 .NET 运行时和所有依赖
        var baseDir = AppContext.BaseDirectory;

        // 优先查找 ClaudeLight.CLI.exe（独立的 CLI 工具）
        var cliExe = Path.Combine(baseDir, "ClaudeLight.CLI.exe");

        if (File.Exists(cliExe))
            return cliExe;

        // 如果找不到 CLI，返回空路径（会在 InstallHooks 中处理）
        return string.Empty;
    }

    public static void InstallHooks(string settingsPath, string cliExePath)
    {
        var json = File.Exists(settingsPath) ? File.ReadAllText(settingsPath) : "{}";
        var doc = JsonDocument.Parse(json);
        var root = doc.RootElement.Clone();

        // Check if claude-light hooks are already installed
        if (root.TryGetProperty("hooks", out var hooks) &&
            hooks.GetRawText().Contains(HookMarker))
        {
            return;
        }

        var cliPath = cliExePath.Replace("\\", "\\\\");

        // Parse existing hooks (excluding claude-light hooks)
        var existingHooksJson = new Dictionary<string, string>();
        if (root.TryGetProperty("hooks", out var existingHooksProp))
        {
            foreach (var prop in existingHooksProp.EnumerateObject())
            {
                // Skip claude-light hooks
                if (prop.Value.GetRawText().Contains(HookMarker))
                    continue;
                existingHooksJson[prop.Name] = prop.Value.GetRawText();
            }
        }

        // Build final hooks JSON manually
        var hooksContent = "{";
        var first = true;

        // Add existing hooks
        foreach (var kvp in existingHooksJson)
        {
            if (!first) hooksContent += ",";
            hooksContent += $"\"{kvp.Key}\":{kvp.Value}";
            first = false;
        }

        // Add claude-light hooks
        if (!first) hooksContent += ",";
        hooksContent += $@"
  ""PreToolUse"": [{{ ""hooks"": [{{ ""type"": ""command"", ""command"": ""\""{cliPath}\"" hook running $CLAUDE_PROJECT_DIR"" }}] }}],
  ""PostToolUse"": [{{ ""hooks"": [{{ ""type"": ""command"", ""command"": ""\""{cliPath}\"" hook done $CLAUDE_PROJECT_DIR"" }}] }}],
  ""Notification"": [{{ ""hooks"": [{{ ""type"": ""command"", ""command"": ""\""{cliPath}\"" hook confirm $CLAUDE_PROJECT_DIR"" }}] }}]
}}";

        // Rebuild the entire settings object
        var result = new Dictionary<string, string>();
        foreach (var prop in root.EnumerateObject())
        {
            if (prop.Name == "hooks")
                continue; // Skip old hooks, will add merged version
            result[prop.Name] = prop.Value.GetRawText();
        }
        result["hooks"] = hooksContent;

        // Build final JSON
        var finalJson = "{";
        var firstProp = true;
        foreach (var kvp in result)
        {
            if (!firstProp) finalJson += ",";
            finalJson += $"\"{kvp.Key}\":{kvp.Value}";
            firstProp = false;
        }
        finalJson += "}";

        // Re-parse and re-serialize for proper formatting
        var finalDoc = JsonDocument.Parse(finalJson);
        var merged = JsonSerializer.Serialize(finalDoc, new JsonSerializerOptions { WriteIndented = true });
        File.WriteAllText(settingsPath, merged);
    }

    public static void RemoveHooks(string settingsPath)
    {
        if (!File.Exists(settingsPath)) return;

        var json = File.ReadAllText(settingsPath);
        if (!json.Contains(HookMarker)) return;

        var doc = JsonDocument.Parse(json);
        var root = doc.RootElement.Clone();

        if (!root.TryGetProperty("hooks", out var hooksProp))
            return;

        // Keep only non-claude-light hooks
        var remainingHooks = new Dictionary<string, string>();
        foreach (var prop in hooksProp.EnumerateObject())
        {
            // Skip claude-light hooks
            if (prop.Value.GetRawText().Contains(HookMarker))
                continue;
            remainingHooks[prop.Name] = prop.Value.GetRawText();
        }

        // Rebuild settings without claude-light hooks
        var result = new Dictionary<string, string>();
        foreach (var prop in root.EnumerateObject())
        {
            if (prop.Name == "hooks")
                continue;
            result[prop.Name] = prop.Value.GetRawText();
        }

        // Add remaining hooks if any
        if (remainingHooks.Count > 0)
        {
            var hooksContent = "{";
            var first = true;
            foreach (var kvp in remainingHooks)
            {
                if (!first) hooksContent += ",";
                hooksContent += $"\"{kvp.Key}\":{kvp.Value}";
                first = false;
            }
            hooksContent += "}";
            result["hooks"] = hooksContent;
        }

        // Build final JSON
        var finalJson = "{";
        var firstProp = true;
        foreach (var kvp in result)
        {
            if (!firstProp) finalJson += ",";
            finalJson += $"\"{kvp.Key}\":{kvp.Value}";
            firstProp = false;
        }
        finalJson += "}";

        // Re-parse and re-serialize for proper formatting
        var finalDoc = JsonDocument.Parse(finalJson);
        var cleaned = JsonSerializer.Serialize(finalDoc, new JsonSerializerOptions { WriteIndented = true });
        File.WriteAllText(settingsPath, cleaned);
    }
}
